saleor/checkout/complete_checkout.py

```diff
@@ -115,14 +115,50 @@ def _process_voucher_data_for_order(checkout_info: "CheckoutInfo") -> dict:
         return {}
 
     customer_email = cast(str, checkout_info.get_customer_email())
-    increase_voucher_usage(voucher, voucher_code, customer_email)
 
+    _increase_checkout_voucher_usage(checkout, voucher_code, voucher, customer_email)
     return {
         "voucher": voucher,
         "voucher_code": voucher_code.code,
     }
 
 
+@traced_atomic_transaction()
+def _increase_checkout_voucher_usage(
+    checkout: "Checkout",
+    voucher_code: "VoucherCode",
+    voucher: "Voucher",
+    customer_email: str,
+):
+    # Prevent race condition when two different threads are processing the same checkout
+    # with limited usage voucher assigned, both threads increasing the
+    # voucher usage which causing `NotApplicable` error for voucher.
+    if checkout.is_voucher_usage_increased:
+        return
+    increase_voucher_usage(voucher, voucher_code, customer_email)
+    checkout.is_voucher_usage_increased = True
+    checkout.save(update_fields=["is_voucher_usage_increased"])
+
+
+@traced_atomic_transaction()
+def _release_checkout_voucher_usage(
+    checkout: "Checkout",
+    voucher_code: Optional["VoucherCode"],
+    voucher: Optional["Voucher"],
+    user_email: Optional[str],
+):
+    if not checkout.is_voucher_usage_increased or not voucher_code:
```

Comment:
Shouldn't be a separate condition?
It will exit when checkout.is_voucher_usage_increased set to True, but no `voucher_code`. 
Shoudln't we have something like:
```python
if checkout.is_voucher_usage_increated:
    checkout.is_voucher_usage_increated = False
    checkout.save(..)
if not voucher_code:
    return

(rest of the logic)
```

Reply:
Good point, and I forgot to change the `is_voucher_usage_increased` when the voucher is removed from the checkout.